src/utils/pathCheck.ts

```diff
@@ -0,0 +1,112 @@
+import fs from 'fs';
+import path from 'path';
+
+import semver from 'semver';
+import * as childProcess from '@/utils/childProcess';
+import resources from '@/resources';
+
+const flags: Record<string, string> = {
+  helm:    'version',
+  kim:     '-v',
+  kubectl: 'version',
+};
+const regexes: Record<string, RegExp> = {
+  // helm has to match both
+  // current: version.BuildInfo{Version:"v3.5.3", ...
+  // older:   Client: &version.Version{SemVer:"v2.16.12", ...
+  helm:    /Version.*:.*?"v(.+?)"/,
+  kim:     /version v(\S+)/,
+  kubectl: /Client Version.*?GitVersion:"v(.+?)"/,
+};
+
+export default async function shadowInfo(targetDir: string, binaryName: string): Promise<Array<string>> {
+  const notes: Array<string> = [];
+  const referencePath = resources.executable(binaryName);
+
+  try {
+    await fs.promises.access(referencePath, fs.constants.R_OK | fs.constants.X_OK);
+  } catch (err) {
+    if (err.code === 'ENOENT') {
+      console.log(err);
+
+      return notes;
+    }
+  }
+  const proposedVersion = await getVersion(referencePath, binaryName);
+
+  if (!proposedVersion) {
+    return notes;
+  }
+  const paths: Array<string> = process.env.PATH?.split(path.delimiter) ?? [];
+
+  let sawCurrentDir = false;
+
+  for (const currentDir of paths) {
+    if (currentDir === targetDir) {
+      sawCurrentDir = true;
+      continue;
+    }
+    const currentPath = path.join(currentDir, binaryName);
+
+    try {
+      await fs.promises.access(currentPath, fs.constants.X_OK);
+    } catch (err) {
+      continue;
+    }
+
+    // For kubectl, don't bother comparing versions, just existence is enough of a problem
+    // if it occurs earlier in the path, because our kubectl is actually a symlink to kuberlr
+    if (binaryName === 'kubectl') {
+      if (!sawCurrentDir) {
+        notes.push(`Existing instance of ${ binaryName } in ${ currentDir } hinders internal linking of kubectl to kuberlr.`);
+      }
+      continue;
+    }
+    const currentVersion = await getVersion(currentPath, binaryName);
+
+    if (!currentVersion) {
+      // If the tested executable gives unexpected output, ignore it -- it could be
+      // due to any problem, such as copying /bin/ls into a directory above
+      // /usr/local/bin/ and calling the copy `kim`. We can't catch all those problems.
+      continue;
+    }
+
+    // complain about all earlier instances in the path if the version is different
+    // complain about later instances only if they're newer
+    if (!sawCurrentDir) {
+      if (currentVersion.compare(proposedVersion) !== 0) {
```

Comment:
```suggestion
      if (!currentVersion.eq(proposedVersion)) {
```
This is the same, but is slightly clearer that we're checking equality here.

Reply:
(From discussion: this was incorrect, there's no `SemVer.prototype.eq`, etc; they're functions on the semver package, which doesn't look quite as nice.)